### PR TITLE
[Engine] Add no op surface

### PR DESCRIPTION
**Pull request by jonahwilliams**
Still needs to be tested in metalless environment. This should probably print out some kind of error message that tells folks that this is WAI, but I'm not sure if that would be too disruptive for g3.

Fixes https://github.com/flutter/flutter/issues/153883

### CODE REVIEW

File: shell/gpu/gpu_surface_noop.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#define FLUTTER_SHELL_GPU_GPU_SURFACE_NOOP_H_
+
+#include <Metal/Metal.h>
+
+#include "flutter/flow/surface.h"
+#include "flutter/fml/macros.h"
+
+namespace flutter {
+
+class GPUSurfaceNoop : public Surface {
```
Comment (chinmaygarde) [inline shell/gpu/gpu_surface_noop.h:15]: A short docstring saying something like "Accepts rendering intent but doesn't render anything. Useful for tests that don't care about checking rendering output but still need to run a Flutter app." Or something like that.

File: shell/gpu/gpu_surface_noop.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  // |Surface|
+  bool AllowsDrawingWhenGpuDisabled() const override;
+
+  // |Surface|
+  bool EnableRasterCache() const override;
+
+  // |Surface|
+  std::shared_ptr<impeller::AiksContext> GetAiksContext() const override;
+
+  FML_DISALLOW_COPY_AND_ASSIGN(GPUSurfaceNoop);
```
Comment (chinmaygarde) [inline shell/gpu/gpu_surface_noop.h:57]: Here and elsewhere, avoid using the macros.

**Review by chinmaygarde — APPROVED** [review]
The creation of the context happens in https://github.com/flutter/engine/blob/cf4507ae1008c11c6155d31e852999d75670afcf/impeller/renderer/backend/metal/context_mtl.mm#L209. Perhaps (in simulator builds only) we can add a hook for tests to return nullptr from there to test the fallback logic?

**Comment by jonahwilliams** [comment]
I don't think overriding it at that location would be sufficient, nor would a unit test cover it.

**Comment by jonahwilliams** [comment]
I think no matter what we do, we can't really this will be certain for all of the various engine APIS. for example, what happens if the noop surface is asked to use the snapshot controller? or upload an image. Today we may count on that falling back to software.

I'd rather not spin up an entire new test shard for this. Maybe we land and then explore the impact in g3?

File: shell/gpu/gpu_surface_noop.h (hunk 1 above)
Comment (jonahwilliams) [reply shell/gpu/gpu_surface_noop.h:19]: Done

File: shell/gpu/gpu_surface_noop.h (hunk 2 above)
Comment (jonahwilliams) [reply shell/gpu/gpu_surface_noop.h]: Done

**Pull request merged**